sci-log-db/src/mixins/basesnippet.repository-mixin.ts

```diff
@@ -0,0 +1,378 @@
+      basesnippet: Basesnippet,
+      user: UserProfile,
+    ): Promise<void> {
+      if (
+        typeof basesnippet.deleted == 'undefined' ||
+        basesnippet.deleted === false
+      ) {
+        const snippet = await this.findById(id, {}, {currentUser: user});
+        if (
+          (typeof snippet?.expiresAt != 'undefined' &&
+            snippet.expiresAt.getTime() < Date.now()) ||
+          typeof snippet?.expiresAt == 'undefined'
+        ) {
+          throw new HttpErrors.Forbidden('Cannot modify expired data snippet.');
+        }
+        if (snippet?.versionable) {
+          await this.addToHistory(snippet, user);
+        }
+      }
+      await this.updateById(id, basesnippet, {
+        currentUser: user,
+      });
+    }
+
+    private async addToHistory(snippet: M, user: UserProfile): Promise<void> {
+      const historySnippet = await this.getHistorySnippet(snippet, user);
+      const snippetCopy = _.omit(snippet, 'id');
+      snippetCopy.parentId = historySnippet.id;
+      const baseSnippetRepository = await this.baseSnippetRepository();
+      await baseSnippetRepository.create(snippetCopy, {currentUser: user});
+    }
+
+    private async getHistorySnippet(
+      snippet: M,
+      user: UserProfile,
+    ): Promise<Basesnippet> {
+      const baseSnippetRepository = await this.baseSnippetRepository();
+      let historySnippet = await baseSnippetRepository.findOne(
+        {where: {snippetType: 'history', parentId: snippet.id}},
+        {currentUser: user},
+      );
+
+      if (historySnippet == null) {
+        const historySnippetPayload = _.pick(snippet, [
+          'isPrivate',
+          'ownerGroup',
+          'accessGroups',
+        ]);
+        historySnippetPayload.parentId = snippet.id;
+        historySnippetPayload.snippetType = 'history';
+        historySnippet = await baseSnippetRepository.create(
+          historySnippetPayload,
+          {currentUser: user},
+        );
+      }
+      return historySnippet;
+    }
+
+    async deleteByIdWithHistory(id: ID, user: UserProfile): Promise<void> {
+      // Two steps:
+      // 1. set snippet to 'deleted=true'
+      // 2. inside websocket and after informing the clients, replace the parentId or delete the snippet
+      const snippet = await this.findById(id, {}, {currentUser: user});
+      if (snippet?.versionable) {
+        if (snippet?.parentId) {
+          const baseSnippetRepository = await this.baseSnippetRepository();
+          const parent = await baseSnippetRepository.findById(
+            (snippet.parentId as unknown) as ID,
+            {},
+            {currentUser: user},
+          );
+          const parentHistory = await this.getHistorySnippet(parent, user);
+          console.log('deleteById:parentHistory:', parentHistory);
+        }
+      }
+      await this.updateById(id as ID, {deleted: true}, {currentUser: user});
+    }
+
+    async restoreDeletedId(id: ID, user: UserProfile): Promise<void> {
```

Comment:
here, following the comment above, if on delete we created a copy of the original snippet, updated the original one (by setting deleted: true) and created the history snippet (basically what happens on updateWithHistory), then restoring would simply mean updateWithHistory the updated one

Reply:
yes, sounds good

Reply:
I'll keep this change for a later PR